cyberbrain/value_stack.py

```diff
@@ -88,16 +99,83 @@ class Why(enum.Enum):
     SILENCED = 8  # Exception silenced by 'with'
 
 
+class StackItem:
+    """Class representing an item on the value stack.
+    Normal items will have a lineno (starting lineno if multiline) and a list of sources and is_custom will be false.
+    Custom items (Exceptions, Why's, anything not tracked by Cyberbrain) will have the value stored in custom_value.
+    """
+
+    def __init__(
+        self,
+        start_lineno: int,
+        sources: List[Symbol],
+        custom_value: any = None,
+        is_custom: bool = False,
+    ):
+        self.start_lineno = start_lineno
+        self.sources = sources
+        self.custom_value = custom_value
+        self.is_custom = is_custom
+
+    def __repr__(self) -> str:
+        if self.is_custom:
+            return repr(self.custom_value)
+        return "{" + repr(self.start_lineno) + " " + repr(self.sources) + "}"
+
+    def get_sources(self) -> List[Symbol]:
+        if self.is_custom:
+            return []
+        return self.sources
+
+    def get_top_source(self) -> Symbol:
+        if self.is_custom or len(self.sources) == 0:
+            return None
+        return self.sources[0]
+
+    def copy(self) -> StackItem:
+        if self.is_custom:
+            return StackItem(self.start_lineno, [], self.custom_value, self.is_custom)
+        return StackItem(
+            self.start_lineno, [copy(sym) for sym in self.sources], self.custom_value
+        )
+
+    def is_custom_none(self) -> bool:
+        return (self.is_custom and self.custom_value is None) or (
+            not self.is_custom and self.custom_value is _none_placeholder
+        )
+
+
+def merge_items(*values: List[StackItem]) -> StackItem:
```

Comment:
instead of having a free function, I'd suggest creating a factory function in `class StackItem`, something like

```
@static_method
def from_stack_items(*stack_items: List[StackItem]) -> StackItem:
```